mm0-rs/src/elab/lisp/debug.rs

```diff
@@ -0,0 +1,1465 @@
+//! Debug formatting for mm0-rs items. Follows the API set out by the `print` and `pretty`
+//! modules. 
+use std::collections::HashMap;
+use std::cell::RefCell;
+use std::{fmt::{self}};
+use shoebill::{ Printer, Doclike, DocPtr };
+use shoebill::ron::{ RonStruct, RonSequence, RonTuple, RonOption, RonResult };
+use super::{LispVal, LispKind, print::FormatEnv,
+  super::{
+    environment::{DeclKey, NotaInfo, AtomData, AtomID, SortID, TermID, ThmID, Term, Thm, Type, Sort, Expr }, }};
+use crate::util::{ FileSpan, FileRef, Span };
+
+/// Debug counterpart to the `Print` struct`.
+pub struct DebugPrint<'a, D : ?Sized> {
+  fe : FormatEnv<'a>,
+  e : &'a D
+}
+
+/// DebugPrint items can be put in formatters using `{:?}`
+impl<'a, D: EnvDebug + ?Sized> fmt::Debug for DebugPrint<'a, D> {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result { self.e.dbg(self.fe, f) }
+}
+
+impl<'a> FormatEnv<'a> {
+  /// The debug counterpart to `FormatEnv::pretty()`
+  pub fn debug<T>(self, f : impl for<'b> FnOnce(FormatEnv<'b>, &'b mut Printer<'b>) -> T) -> T {
+    f(self, &mut Printer::new())
+  }
+
+  /// The debug counterpart to `FormatEnv::to()`
+  pub fn to_debug<D : ?Sized>(self, e : &'a D) -> DebugPrint<'a, D> {
+    DebugPrint {fe : self, e}
+  }
+}
+
+impl EnvDebug for NotaInfo {
+  fn dbg_aux<'a>(
+    &self, 
+    fe : FormatEnv<'a>,
+    p : &mut Printer<'a>, 
+  ) -> DocPtr<'a> {
+      let mut ron = RonStruct::new();
+      ron.add_field("span", self.span.dbg_aux(fe, p));
+      ron.add_field("term", self.term.dbg_aux(fe, p));
+      ron.add_field("nargs", self.nargs.dbg_aux(fe, p));
+      ron.add_field("rassoc", self.rassoc.dbg_aux(fe, p));
+      ron.add_field("lits", self.lits.dbg_aux(fe, p));
+      ron.to_doc(p)
+  }
+}
+
+
+impl EnvDebug for DeclKey {
+  fn dbg_aux<'a>(
+    &self, 
+    fe : FormatEnv<'a>,
+    p : &mut Printer<'a>, 
+  ) -> DocPtr<'a> {
+      let mut ron = RonTuple::new();
+      ron.add_name("DeclKey");
+      match self {
+          DeclKey::Term(term_id) => {
+              ron.add_name("Term");
+              ron.add_field(term_id.dbg_aux(fe, p));
+          },
+          DeclKey::Thm(thm_id) => {
+              ron.add_name("Thm");
+              ron.add_field(thm_id.dbg_aux(fe, p));
+          }
+      }
+      ron.to_doc(p)
+  }
+}
+
+impl EnvDebug for SortID {
+  fn dbg_aux<'a>(
+    &self, 
+    fe : FormatEnv<'a>,
+    p : &mut Printer<'a>, 
+  ) -> DocPtr<'a> {
+    match self {
+      SortID(n) => {
+        let sort = &fe.sorts[*self];
+        let mut ron = RonStruct::new();
+        ron.add_name("Sort");
+        ron.add_field("sort_id", n.to_string());
```

Comment:
Matching on the `SortID(n)` here is a bit odd, since the whole API uses `SortID` directly. From a debugging point of view I would think you would only want to show names instead of the underlying numbers, unless you need to cross reference with something else that uses numbers. But printing the full sort data whenever you see a reference to it seems like overkill and may even lead to infinite recursion (as noted). I would suggest just printing the atom name, i.e. `fe.sorts[*self].name`.